libbeat/publisher/queue/spool/codec.go

```diff
@@ -0,0 +1,182 @@
+package spool
+
+import (
+	"bytes"
+	"fmt"
+	"time"
+
+	"github.com/elastic/go-structform"
+	"github.com/elastic/go-structform/cborl"
+	"github.com/elastic/go-structform/gotype"
+	"github.com/elastic/go-structform/json"
+	"github.com/elastic/go-structform/ubjson"
+
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/publisher"
+)
+
+type encoder struct {
+	buf    bytes.Buffer
+	folder *gotype.Iterator
+	codec  codecID
+}
+
+type decoder struct {
+	// TODO: replace with more intelligent custom buffer, so to limit amount of memory the buffer
+	//       holds onto after reading a 'large' event.
+	//       Alternative use strucform decoder types (with shared io.Buffer).
+	buf []byte
+
+	json     *json.Parser
+	cborl    *cborl.Parser
+	ubjson   *ubjson.Parser
+	unfolder *gotype.Unfolder
+}
+
+type codecID uint8
+
+type entry struct {
+	Timestamp int64
+	Flags     uint8
+	Meta      common.MapStr
+	Fields    common.MapStr
+}
+
+const (
+	// Note: Never change order. Codec IDs must be not change in the future. Only
+	//       adding new IDs is allowed.
+	codecUnknown codecID = iota
```

Comment:
Not sure I understand why we need different codes. It seems it could be configurable through the config but it's not listed in the reference file. I assume it would also cause issues if one changes the code after the beats was running. 

It's good for testing to have the different codes but should we for now just set one?

Reply:
The JSON codec is nice for debugging(assuming I would have to use an hexeditor). Checking out a file in an hexeditor can be complicated enough. No need to make debugging more complicated by having binary data only.

The UBJSON can be removed.

Benchmarks are available in the go-structform tests (using beats events):

```
BenchmarkDecodeBeatsEvents/packetbeat/structform-json-4         	     200	   9210052 ns/op	  74.12 MB/s	 3251196 B/op	   62062 allocs/op
BenchmarkDecodeBeatsEvents/packetbeat/structform-ubjson-4       	     200	   7440986 ns/op	  87.19 MB/s	 3171845 B/op	   56362 allocs/op
BenchmarkDecodeBeatsEvents/packetbeat/structform-cborl-4        	     200	   7310081 ns/op	  81.46 MB/s	 3171266 B/op	   56361 allocs/op
BenchmarkDecodeBeatsEvents/metricbeat/structform-json-4         	     500	   3333191 ns/op	  62.41 MB/s	 1387710 B/op	   24852 allocs/op
BenchmarkDecodeBeatsEvents/metricbeat/structform-ubjson-4       	     500	   2955721 ns/op	  72.92 MB/s	 1387472 B/op	   24758 allocs/op
BenchmarkDecodeBeatsEvents/metricbeat/structform-cborl-4        	     500	   2816873 ns/op	  67.70 MB/s	 1387136 B/op	   24758 allocs/op
BenchmarkDecodeBeatsEvents/filebeat/structform-json-4           	      30	  46966885 ns/op	 105.49 MB/s	15055009 B/op	  390517 allocs/op
BenchmarkDecodeBeatsEvents/filebeat/structform-ubjson-4         	      30	  38593911 ns/op	 126.92 MB/s	14874314 B/op	  390499 allocs/op
BenchmarkDecodeBeatsEvents/filebeat/structform-cborl-4          	      50	  36702929 ns/op	 123.75 MB/s	14874551 B/op	  390500 allocs/op
BenchmarkEncodeBeatsEvents/packetbeat/structform-json-4         	     200	   5970464 ns/op	 114.26 MB/s	       0 B/op	       0 allocs/op
BenchmarkEncodeBeatsEvents/packetbeat/structform-ubjson-4       	     500	   3749375 ns/op	 173.04 MB/s	       0 B/op	       0 allocs/op
BenchmarkEncodeBeatsEvents/packetbeat/structform-cborl-4        	     500	   2829715 ns/op	 210.43 MB/s	       0 B/op	       0 allocs/op
BenchmarkEncodeBeatsEvents/metricbeat/structform-json-4         	    1000	   2274028 ns/op	  91.03 MB/s	       0 B/op	       0 allocs/op
BenchmarkEncodeBeatsEvents/metricbeat/structform-ubjson-4       	    1000	   1672348 ns/op	 128.88 MB/s	       0 B/op	       0 allocs/op
BenchmarkEncodeBeatsEvents/metricbeat/structform-cborl-4        	    1000	   1236829 ns/op	 154.18 MB/s	       0 B/op	       0 allocs/op
BenchmarkEncodeBeatsEvents/filebeat/structform-json-4           	      50	  32523323 ns/op	 151.97 MB/s	       0 B/op	       0 allocs/op
BenchmarkEncodeBeatsEvents/filebeat/structform-ubjson-4         	     100	  22533875 ns/op	 217.37 MB/s	       0 B/op	       0 allocs/op
BenchmarkEncodeBeatsEvents/filebeat/structform-cborl-4          	     100	  15919294 ns/op	 285.32 MB/s	       0 B/op	       0 allocs/op
```

ubjson is faster on decode, cborl is much faster on encode :)

Reply:
btw. the codec setting is not documented for now.

Reply:
For not documenting it: If the config is there, it should be in the reference file. We did this "trick" in the past and I think it didn't benefit us as people discovered the option and started using it. Better have it in the reference file as experimental and say what it does.

If I understand the comment from ph a user can have code A and then configure codec B on restart and everything will keep working smootly? The events are read in with codec A and the next time the file is written, it will be all in codec B.

Reply:
Added `codec` setting to the reference file.

> If I understand the comment from ph a user can have code A and then configure codec B on restart and everything will keep working smootly? The events are read in with codec A and the next time the file is written, it will be all in codec B.

The codec can be change between restarts. The setting is only used by the write. The reader can easily deal with all codecs within a file.